### PR TITLE
feat: implement KubeSpan manager for Wireguard peer state

### PR DESCRIPTION
**Pull request by smira**
KubeSpan manager uses list of KubeSpan peers prepared from the discovery
and local KubeSpan identity to set up and update configuration of the
Wireguard interface.

As new peers are getting added or deleted, manager takes care of
updating the Wireguard config.

Manager also keeps track of all peers and their state coming from the
Wireguard link status: whether the connection is up or not, some stats,
last actually used endpoint, etc.

Manager cycles through the available peer endpoints until it finds the
one which works.

Manager exposes peer status as `PeerStatus` resources.

Example:

```
$ talosctl -n 172.20.0.2 get kubespanpeerstatuses
NODE         NAMESPACE   TYPE                 ID                                             VERSION   LABEL                    ENDPOINT           STATE   RX    TX
172.20.0.2   kubespan    KubeSpanPeerStatus   GpO3gs5n09WpoiVANbzRL5nwrkRi+9Q19qoeC8RTkQ4=   30        talos-default-worker-2   172.20.0.6:51820   up      640   1920
172.20.0.2   kubespan    KubeSpanPeerStatus   j4CRlKByMcTWOBS2ifZcPzcUr3lXdBOc/I4AxGmhXxI=   30        talos-default-worker-1   172.20.0.5:51820   up      672   1888
172.20.0.2   kubespan    KubeSpanPeerStatus   o5EPScFrD895A5EpVyKU8hFR+vi25D0CJMYsoaXN3Qk=   28        talos-default-master-3   172.20.0.4:51820   up      640   1920
172.20.0.2   kubespan    KubeSpanPeerStatus   rBp5wyHdxqZkq5CWher2DcPcGgwHrFOwB6fP/ReFRlE=   16        talos-default-master-2   172.20.0.3:51820   up      432   2088
```

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>
Signed-off-by: Seán C McCord <ulexus@gmail.com>
Co-authored-by: Seán C McCord <ulexus@gmail.com>

### CODE REVIEW

File: internal/app/machined/pkg/controllers/cluster/local_affiliate.go [hunk 1]
```diff
@@ -197,6 +197,8 @@ func (ctrl *LocalAffiliateController) Run(ctx context.Context, r controller.Runt
 						}
 
 						spec.KubeSpan.Endpoints = endpoints
+					} else {
+						spec.KubeSpan = cluster.KubeSpanAffiliateSpec{}
```
Comment (Ulexus) [inline internal/app/machined/pkg/controllers/cluster/local_affiliate.go:201]: I'm not sure I follow this thing, but can we not just do a `if spec.KubeSpan == nil { spec.KubeSpan = cluster.KubeSpanAffiliateSpec{} }` _before_ this `if` and thus eliminate the `else`?

Comment (Ulexus) [reply internal/app/machined/pkg/controllers/cluster/local_affiliate.go:201]: #readability

File: internal/app/machined/pkg/controllers/kubespan/manager.go [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+		cfgSpec := cfg.(*kubespan.Config).TypedSpec()
+
+		localIdentity, err := r.Get(ctx, resource.NewMetadata(kubespan.NamespaceName, kubespan.IdentityType, kubespan.LocalIdentity, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting first MAC address: %w", err)
```
Comment (Ulexus) [inline internal/app/machined/pkg/controllers/kubespan/manager.go:170]: `localIdentity !== first mac`?

File: pkg/machinery/constants/constants.go [hunk 3]
```diff
@@ -474,6 +474,25 @@ const (
...
 	KubeSpanDefaultPort = 51820
 
+	// KubeSpanDefaultRoutingTable is the default routing table for KubeSpan LAN targets.
+	//
+	// This specifies the routing table which will be used for Wireguard-available destinations.
+	KubeSpanDefaultRoutingTable = 180
+
+	// KubeSpanDefaultFirewallMark is the default firewall mark to use for Wireguard encrypted egress packets.
+	//
+	// The next consecutive mark will be used by KubeSpan to mark packets which should be forced into the Wireguard interface.
```
Comment (Ulexus) [inline pkg/machinery/constants/constants.go:484]: This comment it now outdated:  we are defining the force port explicitly next; not using the next-consecutive necessarily.

File: internal/app/machined/pkg/controllers/kubespan/manager.go (hunk 2 above)
Comment (Ulexus) [reply internal/app/machined/pkg/controllers/kubespan/manager.go:170]: I mean, it _is_ "currently" the MAC, but the local code does not match that... it's simply the identity

File: internal/app/machined/pkg/controllers/cluster/local_affiliate.go (hunk 1 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/cluster/local_affiliate.go:201]: This one is simple - Affiliate contains current state for the local node. If KubeSpan was enabled, Affiliate might have had KubeSpan information. But on this reconcile loop we see that KubeSpan is disabled, so we need to clear KubeSpan settings of the Affiliate.

We don't need to clear it in general

File: internal/app/machined/pkg/controllers/kubespan/manager.go (hunk 2 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/kubespan/manager.go:170]: That's just wrong error message, copy-paste, I'll fix it, thanks

File: pkg/machinery/constants/constants.go (hunk 3 above)
Comment (smira) [reply pkg/machinery/constants/constants.go:484]: Ok, I will remove that

File: internal/app/machined/pkg/controllers/cluster/local_affiliate.go (hunk 1 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/cluster/local_affiliate.go:201]: but I got your point, I'll update it, thanks

**Review by Ulexus — APPROVED** [review]
> Reviewed 17 of 17 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/talos-systems/talos/4211)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @smira)

**Pull request merged**